dot/parachain/types/async_backing.go

```diff
@@ -17,3 +17,76 @@ type AsyncBackingParams struct {
 	AllowedAncestryLen uint32 `scale:"2"`
 }
+
+// InboundHrmpLimitations constraints on inbound HRMP channels.
+type InboundHrmpLimitations struct {
+	// An exhaustive set of all valid watermarks, sorted ascending.
+	//
+	// It's only expected to contain block numbers at which messages were
+	// previously sent to a para, excluding most recent head.
+	ValidWatermarks []BlockNumber
+}
+
+// OutboundHrmpChannelLimitations constraints on outbound HRMP channels.
+type OutboundHrmpChannelLimitations struct {
+	// The maximum bytes that can be written to the channel.
+	BytesRemaining uint32
+	// The maximum messages that can be written to the channel.
+	MessagesRemaining uint32
+}
+
+// Constraints on the actions that can be taken by a new parachain block. These
+// limitations are implicitly associated with some particular parachain, which should
+// be apparent from usage.
+type Constraints struct {
+	// The minimum relay-parent number accepted under these constraints.
+	MinRelayParentNumber BlockNumber
+	// The maximum Proof-of-Validity size allowed, in bytes.
+	MaxPoVSize uint32
+	// The maximum new validation code size allowed, in bytes.
+	MaxCodeSize uint32
+	// The amount of UMP messages remaining.
+	UmpRemaining uint32
+	// The amount of UMP bytes remaining.
+	UmpRemainingBytes uint32
+	// The maximum number of UMP messages allowed per candidate.
+	MaxUmpNumPerCandidate uint32
+	// Remaining DMP queue. Only includes sent-at block numbers.
+	DmpRemainingMessages []uint32
+	// The limitations of all registered inbound HRMP channels.
+	HrmpInbound InboundHrmpLimitations
+	// The limitations of all registered outbound HRMP channels.
+	HrmpChannelsOut map[ParaID]OutboundHrmpChannelLimitations
+	// The maximum number of HRMP messages allowed per candidate.
+	MaxHrmpNumPerCandidate uint32
+	// The required parent head-data of the parachain.
+	RequiredParent HeadData
+	// The expected validation-code-hash of this parachain.
+	ValidationCodeHash ValidationCodeHash
+	// The code upgrade restriction signal as-of this parachain.
+	UpgradeRestriction *UpgradeRestriction
+	// The future validation code hash, if any, and at what relay-parent
+	// number the upgrade would be minimally applied.
+	FutureValidationCode *FutureValidationCode
+}
+
+// FutureValidationCode represents a tuple of BlockNumber and ValidationCodeHash
+type FutureValidationCode struct {
+	BlockNumber        BlockNumber
+	ValidationCodeHash ValidationCodeHash
+}
+
+type CandidatePendingAvailability struct {
+	CandidateHash     CandidateHash
+	Descriptor        CandidateDescriptorV2
+	Commitments       CandidateCommitments
+	RelayParentNumber BlockNumber
+	MaxPoVSize        uint32
+}
+
+// BackingState holds the state of the backing system per-parachain, including
+// state-machine constraints and candidates pending availability
+type BackingState struct {
+	Constraints         Constraints
+	PendingAvailability []CandidatePendingAvailability
+}
```

Comment:
It would be nice if you could add a scale tag to the fields of this struct and all the related structs that need to be encoded/decoded properly.

Reply:
I can place, but they will are useless, I would recommend only when we face types that use the `#[codec(index=..)]` directive on top of the types, eg. https://github.com/paritytech/polkadot-sdk/blob/53f6473c9c8c9d18b5ef0ed02a587757495d1dbf/polkadot/xcm/src/lib.rs#L219

Reply:
Looking at our structure, it is not evident that it needs to be scale encoded/decoded properly.
Someone in future might change the order. It would be nice to have a scale tag.

If you choose not to add a scale tag, I am fine with that, too.

Reply:
> Someone in future might change the order

scale tag still being useless for this case, I prefer to have tests to protect against those modifications, but I've added.